token-erc-1155/chaincode-go/chaincode/contract.go

```diff
@@ -972,47 +975,64 @@ func removeBalance(ctx contractapi.TransactionContextInterface, sender string, i
 				selfRecipientKeyNeedsToBeRemoved = true
 				selfRecipientKey = queryResponse.Key
 			} else {
-				err = ctx.GetStub().DelState(queryResponse.Key)
-				if err != nil {
-					return fmt.Errorf("failed to delete the state of %v: %v", queryResponse.Key, err)
-				}
+				deferredDeletions = append(deferredDeletions, deferredDelete(ctx, queryResponse))
```

Comment:
DelState() won't delete at this time. It will build a write set that will perform the deletes upon successful completion of the chaincode and validation of the transaction. It is therefore implicitly deferred already. Introducing an explicit defer in my opinion makes the chaincode more confusing to the user, and gives the wrong idea that there is some type of explicit deferral possible in the chaincode programming model.

I see your point however about making the implicit deferral more obvious to the user. I'd suggest that the better way to educate the user would be to add a comment at the existing DelState() call indicating that the DelState() calls will get collected in the write set and will only be executed if the remaining checks are passed and the transaction gets validated. This will educate the user on the chaincode programming model and keep the code more readable.

Reply:
This makes sense but when testing with shimtest.MockStub, the change is made at the time its executed and is not part of any writeSet that will ignore the deletion.

A better solution might be to push the deletion further down in the chain of events but I'd like confirmation whether this would be worthwhile/acceptable before I make the change.